filters/openpolicyagent/internal/envoy/skipperadapter_test.go

```diff
@@ -0,0 +1,199 @@
+package envoy
+
+import (
+	ext_authz_v3_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	ext_authz_v3 "github.com/envoyproxy/go-control-plane/envoy/service/auth/v3"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/protobuf/types/known/structpb"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func TestAdaptToExtAuthRequest(t *testing.T) {
+	tests := []struct {
+		name              string
+		req               *http.Request
+		metadata          *ext_authz_v3_core.Metadata
+		contextExtensions map[string]string
+		rawBody           []byte
+		want              *ext_authz_v3.CheckRequest
+		wantErr           bool
+	}{
+		{
+			name: "valid request with headers and metadata",
+			req: &http.Request{
+				Method: "GET",
+				Host:   "example-app",
+				URL:    &url.URL{Path: "/users/profile/amal#segment?param=yes"},
+				Header: createHeaders(map[string]string{
+					"accept":            "*/*",
+					"user-agent":        "curl/7.68.0",
+					"x-request-id":      "1455bbb0-0623-4810-a2c6-df73ffd8863a",
+					"x-forwarded-proto": "http",
+					":authority":        "example-app",
+				}),
+				Proto:         "HTTP/1.1",
+				ContentLength: 100,
+			},
+			metadata: createFilterMetadata(map[string]map[string]string{
+				"envoy.filters.http.header_to_metadata": {"policy_type": "ingress"},
+			}),
+			contextExtensions: map[string]string{
+				"key1": "value1",
+				"key2": "value2",
+			},
+			rawBody: []byte(`{"key":"value"}`),
+			want: &ext_authz_v3.CheckRequest{
+				Attributes: &ext_authz_v3.AttributeContext{
+					Request: &ext_authz_v3.AttributeContext_Request{
+						Http: &ext_authz_v3.AttributeContext_HttpRequest{
+							Host:   "example-app",
+							Method: "GET",
+							Path:   "/users/profile/amal%23segment%3Fparam=yes", //URL encoded
+							Headers: map[string]string{
+								"accept":            "*/*",
+								"user-agent":        "curl/7.68.0",
+								"x-request-id":      "1455bbb0-0623-4810-a2c6-df73ffd8863a",
+								"x-forwarded-proto": "http",
+								":authority":        "example-app",
+							},
+							RawBody: []byte(`{"key":"value"}`),
+						},
+					},
+					ContextExtensions: map[string]string{
+						"key1": "value1",
+						"key2": "value2",
+					},
+					MetadataContext: createFilterMetadata(map[string]map[string]string{
+						"envoy.filters.http.header_to_metadata": {"policy_type": "ingress"},
+					}),
+				},
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := AdaptToExtAuthRequest(tt.req, tt.metadata, tt.contextExtensions, tt.rawBody)
```

Comment:
I think it would make more sense to change the tests to convert this as well to an AST input and validate that one. The Envoy request object itself is sth intermediate. 

Looking forward we could also remove the double serialisation / deserialisation of Skipper Request -> Envoy Protobuf -> JSON -> OPA AST without changing the tests.

Reply:
I agree!
Normally this should be part of some bdd style test with invalid input. If it is too complicated a unit test is fine.